File: Tayyibat/Services/DataExportService.swift
import Foundation
import SwiftData

/// يصدّر بيانات المتابعة إلى ملف JSON قابل للمشاركة.
enum DataExportService {
    private struct Export: Codable {
        var exportedAt: Date
        var meals: [MealExport]
        var summaries: [SummaryExport]
    }
    private struct MealExport: Codable {
        var capturedAt: Date
        var overallScore: Int
        var scoreLabel: String
        var wasEdited: Bool
        var items: [ItemExport]
    }
    private struct ItemExport: Codable {
        var nameAr: String
        var verdict: String
        var category: String
        var confidence: Double
    }
    private struct SummaryExport: Codable {
        var date: Date
        var averageScore: Int
        var mealsCount: Int
        var fastedToday: Bool
    }

    /// ينشئ ملف JSON مؤقتاً ويعيد مساره للمشاركة.
    static func exportFile(context: ModelContext) -> URL? {
        let meals = (try? context.fetch(FetchDescriptor<Meal>(sortBy: [SortDescriptor(\.capturedAt)]))) ?? []
        let summaries = (try? context.fetch(FetchDescriptor<DailySummary>(sortBy: [SortDescriptor(\.date)]))) ?? []

        let export = Export(
            exportedAt: .now,
            meals: meals.map { meal in
                MealExport(
                    capturedAt: meal.capturedAt,
                    overallScore: meal.overallScore,
                    scoreLabel: meal.scoreLabelAr,
                    wasEdited: meal.wasEdited,
                    items: meal.items.map {
                        ItemExport(nameAr: $0.nameAr, verdict: $0.verdictRaw, category: $0.category, confidence: $0.confidence)
                    }
                )
            },
            summaries: summaries.map {
                SummaryExport(date: $0.date, averageScore: $0.averageScore, mealsCount: $0.mealsCount, fastedToday: $0.fastedToday)
            }
        )

        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .withoutEscapingSlashes]
        encoder.dateEncodingStrategy = .iso8601
        guard let data = try? encoder.encode(export) else { return nil }

        let url = FileManager.default.temporaryDirectory.appendingPathComponent("tayyibat_export.json")
        try? data.write(to: url)
        return url
    }

    /// يحذف كل بيانات المتابعة (الوجبات والملخصات وأيام الصيام).
    static func deleteAllTrackingData(context: ModelContext) {
        try? context.delete(model: Meal.self)
        try? context.delete(model: FoodItem.self)
        try? context.delete(model: DailySummary.self)
        try? context.delete(model: FastingDay.self)
        try? context.save()
    }
}

File: android/app/src/main/java/com/tayyibat/app/service/DataExportService.kt
package com.tayyibat.app.service

import android.content.Context
import com.tayyibat.app.data.db.AppDatabase
import com.tayyibat.app.util.DateUtils
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

/** يصدّر بيانات المتابعة إلى ملف JSON قابل للمشاركة. */
object DataExportService {
    private val json = Json { prettyPrint = true }

    @Serializable
    private data class Export(
        val exportedAt: String,
        val meals: List<MealExport>,
        val summaries: List<SummaryExport>,
    )

    @Serializable
    private data class MealExport(
        val capturedAt: String,
        val overallScore: Int,
        val scoreLabel: String,
        val wasEdited: Boolean,
        val items: List<ItemExport>,
    )

    @Serializable
    private data class ItemExport(
        val nameAr: String,
        val verdict: String,
        val category: String,
        val confidence: Double,
    )

    @Serializable
    private data class SummaryExport(
        val date: String,
        val averageScore: Int,
        val mealsCount: Int,
        val fastedToday: Boolean,
    )

    private fun iso(epochMillis: Long): String =
        DateTimeFormatter.ISO_INSTANT.format(Instant.ofEpochMilli(epochMillis).atOffset(ZoneOffset.UTC).toInstant())

    /** ينشئ ملف JSON في ذاكرة التخزين المؤقتة ويعيده للمشاركة. */
    suspend fun exportFile(context: Context, db: AppDatabase): File? = withContext(Dispatchers.IO) {
        val meals = db.mealDao().allMeals()
        val summaries = db.summaryDao().all()

        val export = Export(
            exportedAt = iso(DateUtils.now()),
            meals = meals.map { mw ->
                MealExport(
                    capturedAt = iso(mw.meal.capturedAt),
                    overallScore = mw.meal.overallScore,
                    scoreLabel = mw.meal.scoreLabelAr,
                    wasEdited = mw.meal.wasEdited,
                    items = mw.items.map {
                        ItemExport(it.nameAr, it.verdictRaw, it.category, it.confidence)
                    },
                )
            },
            summaries = summaries.map {
                SummaryExport(iso(it.date), it.averageScore, it.mealsCount, it.fastedToday)
            },
        )

        runCatching {
            val file = File(context.cacheDir, "tayyibat_export.json")
            file.writeText(json.encodeToString(export))
            file
        }.getOrNull()
    }

    /** يحذف كل بيانات المتابعة (الوجبات والملخصات وأيام الصيام). */
    suspend fun deleteAllTrackingData(db: AppDatabase) = withContext(Dispatchers.IO) {
        db.mealDao().deleteAllItems()
        db.mealDao().deleteAllMeals()
        db.summaryDao().deleteAll()
        db.fastingDao().deleteAll()
    }
}
